File: internal/models/deployment_history.go
package models

import "time"

type DeploymentHistory struct {
	ID        string    `json:"id"`
	AppID     string    `json:"app_id"`
	CreatedAt time.Time `json:"created_at"`

	ImageBefore    string `json:"image_before"`
	ImageAfter     string `json:"image_after"`
	ReplicasBefore int    `json:"replicas_before"`
	ReplicasAfter  int    `json:"replicas_after"`

	RequestCPUBefore    int `json:"request_cpu_before"`
	RequestCPUAfter     int `json:"request_cpu_after"`
	RequestMemoryBefore int `json:"request_memory_before"`
	RequestMemoryAfter  int `json:"request_memory_after"`
	LimitCPUBefore      int `json:"limit_cpu_before"`
	LimitCPUAfter       int `json:"limit_cpu_after"`
	LimitMemoryBefore   int `json:"limit_memory_before"`
	LimitMemoryAfter    int `json:"limit_memory_after"`

	DeployType string `json:"deploy_type"`
	DeployedBy string `json:"deployed_by"`
	Reason     string `json:"reason"`
	Status     string `json:"status"`

	BuildID *string `json:"build_id,omitempty"`
}

type RollbackDeploymentRequest struct {
	HistoryID string `json:"history_id" binding:"required"`
}

File: ui/src/api/deployment-history.ts
import client, { type PaginationResponse } from "./client"

export interface DeploymentHistory {
  id: string
  app_id: string
  created_at: string
  
  image_before: string
  image_after: string
  replicas_before: number
  replicas_after: number
  
  request_cpu_before: number
  request_cpu_after: number
  request_memory_before: number
  request_memory_after: number
  limit_cpu_before: number
  limit_cpu_after: number
  limit_memory_before: number
  limit_memory_after: number
  
  deploy_type: string
  deployed_by: string
  reason: string
  status: string
  
  build_id?: string
}

export const deploymentHistoryApi = {
  list: async (appId: string, page = 1, pageSize = 10) => {
    return client.get(`/v1/apps/${appId}/deployment-history`, {
      params: { page, page_size: pageSize }
    }) as Promise<{ items: DeploymentHistory[], pagination: PaginationResponse }>
  },
  
  rollback: async (appId: string, historyId: string) => {
    return client.post(`/v1/apps/${appId}/deployment-history/rollback`, { history_id: historyId })
  }
}
